scripts/scil_denoising_nlmeans.py

```diff
@@ -22,98 +62,173 @@
-    p.add_argument('number_coils', type=int,
-                   help='Number of receiver coils of the scanner.\nUse '
-                        'number_coils=1 in the case of a SENSE (GE, Philips) '
-                        'reconstruction and \nnumber_coils >= 1 for GRAPPA '
-                        'reconstruction (Siemens). number_coils=4 works well '
-                        'for the 1.5T\n in Sherbrooke. Use number_coils=0 if '
-                        'the noise is considered Gaussian distributed.')
-
-    p.add_argument('--mask', metavar='',
-                   help='Path to a binary mask. Only the data inside the mask'
-                        ' will be used for computations')
-    p.add_argument('--sigma', metavar='float', type=float,
-                   help='The standard deviation of the noise to use instead '
-                        'of computing  it automatically.')
-    p.add_argument('--log', dest="logfile",
-                   help='If supplied, name of the text file to store '
-                        'the logs.')
-
+    p.add_argument('--mask_denoise',
+                   help="Path to a binary mask. Only the data inside the mask "
+                        "will be denoised. If not provided, only non-zero "
+                        "voxels will be denoised.")
+    p.add_argument('--gaussian', action='store_true',
+                   help="If you know that your data contains gaussian noise, "
+                        "use this option. Otherwise, Rician is assumed.")
+
+    g = p.add_argument_group("Noise estimation methods")
+    g = g.add_mutually_exclusive_group(required=True)
+    g.add_argument(
+        '--sigma', metavar='float', type=float,
+        help='Provide your own standard deviation of the noise.')
+    g.add_argument(
+        '--basic_sigma', action='store_true',
+        help="Use dipy's basic estimation of sigma.")
+    g.add_argument(
+        '--piesno', action='store_true',
+        help="Estimate sigma using Piesno's method.")
+
+    g = p.add_argument_group("Noise estimation options: piesno and basic")
+    g.add_argument(
+        '--number_coils', type=ranged_type(int, min_value=0),
+        help="Option for Dipy's noise estimation. Here is their description:\n"
+             ">Number of coils of the receiver array. Use N = 1 in case of a\n"
+             ">SENSE reconstruction (Philips scanners) or the number of \n"
+             ">coils for a GRAPPA reconstruction (Siemens and GE). Use 0 to \n"
+             ">disable the correction factor, as for example if the noise is\n"
+             ">Gaussian distributed. See [1] for more information.\n")
+    # ??????????? VOIR SI TOUJOURS D'ACTUALITE????? ON AVAIT: We force to zero as the 3T
+    # is either oversmoothed or still noisy, but we prefer the second option"
+
+    g = p.add_argument_group("Noise estimation options: basic")
+    gg = g.add_mutually_exclusive_group()
+    gg.add_argument(
+        '--mask_sigma',
```

Comment:
On dirait que le mask attendu si l'input est 4D est un mask 4D....? Weird. Possible d'ajuster et supporter un mask 3D en input?

scil_denoising_nlmeans.py dwi.nii.gz dwi_nlm_rician_basic.nii.gz --number_coils 1 --basic_sigma  --mask_sigma piesno_mask.nii.gz --processes 6 -v
Traceback (most recent call last):
  File "/Users/desm2239/my_env/scilpy_master/bin/scil_denoising_nlmeans.py", line 33, in <module>
    sys.exit(load_entry_point('scilpy', 'console_scripts', 'scil_denoising_nlmeans.py')())
  File "/Users/desm2239/Research/Source/scilpy/scripts/scil_denoising_nlmeans.py", line 199, in main
    tmp_vol_data = (vol_data * mask_sigma).astype(np.float32)
ValueError: operands could not be broadcast together with shapes (256,256,62,16) (256,256,62)

Reply:
Thanks! Fixed.